### PR TITLE
Update to use algo object

### LINKED ISSUE #16: Update link.py and mpi.py to use Algorithm object
## Summary
The link.py and mpi.py modules expect a Dict for the algo_config parameter, update these to be an instance of the Algorithm class in models.py.  We may also need to rewrite how the simple_mpi.py and simple_link.py use the algo_config, but only if those modules have been committed to main by the time this issue is worked on.

## Acceptance Criteria
- [ ] Update all references to the algo_config, and its parts, in link.py to use Algorithm
- [ ] Update all references in mpi.py to using parts of an algo_config to using either an Algorithm or AlgorithmPass
- [ ] update any test cases that specify the use of an algo_config Dict to an instance of Algorithm
- [ ] Update the API in main.py to pass in an instance of Algorithm to the `link_record_against_mpi` function

## Dependencies
#15

### PR DESCRIPTION
**Pull request by cbrinson-rise8**
## Description
The link.py and mpi.py modules expect a Dict for the algo_config parameter, update these to be an instance of the Algorithm class in models.py. We may also need to rewrite how the simple_mpi.py and simple_link.py use the algo_config, but only if those modules have been committed to main by the time this issue is worked on.

## Related Issues
closes #16 

## Additional Notes
- Removes uses of the `algo_config` `dict` in the `src.recordlinker.linking` directory and replaces them with uses of `Algorithm` and/or `AlgorithmPass` objects from our database schema

<--------------------- REMOVE THE LINES BELOW BEFORE MERGING --------------------->

## Checklist
Please review and complete the following checklist before submitting your pull request:

- [x] I have ensured that the pull request is of a manageable size, allowing it to be reviewed within a single session.
- [x] I have reviewed my changes to ensure they are clear, concise, and well-documented.
- [x] I have updated the documentation, if applicable.
- [x] I have added or updated test cases to cover my changes, if applicable.
- [x] I have minimized the number of reviewers to include only those essential for the review.
- [x] I have notified teammates in the review thread to build awareness.

## Checklist for Reviewers
Please review and complete the following checklist during the review process:

- [ ] The code follows best practices and conventions.
- [ ] The changes implement the desired functionality or fix the reported issue.
- [ ] The tests cover the new changes and pass successfully.
- [ ] Any potential edge cases or error scenarios have been considered.

### CODE REVIEW

File: src/recordlinker/linking/link.py [hunk 1]
```diff
@@ -82,16 +82,16 @@ def add_person_resource(
...
 
 
-def compare(record: models.PIIRecord, patient: models.Patient, linkage_pass: dict) -> bool:
+def compare(record: models.PIIRecord, patient: models.Patient, algorithm_pass: models.AlgorithmPass) -> bool:
     """
     Compare the incoming record to the linked patient
     """
     # all the functions used for comparison
-    funcs: dict[models.Feature, matchers.FEATURE_COMPARE_FUNC] = linkage_pass["funcs"]
+    funcs: dict[models.Feature, matchers.FEATURE_COMPARE_FUNC] = utils.bind_functions(algorithm_pass.evaluators)
```
Comment (ericbuckley) [inline src/recordlinker/linking/link.py:90]: I was thinking we'd do this a bit earlier.  I think likely when we pull it out of the database, but I haven't really thought that through, so maybe just doing it at the top of link is safest for now.  Anyway, if we have 20 patient records come back, 2 passes and 2 evaluation features to compare, we call this method 80 times.  If we do the binding earlier in the stack, we can reduce the bind_functions overhead.

File: src/recordlinker/linking/link.py [hunk 2]
```diff
@@ -82,16 +82,16 @@ def add_person_resource(
...
     # a function to determine a match based on the comparison results
-    matching_rule: matchers.MATCH_RULE_FUNC = linkage_pass["matching_rule"]
-    # keyword arguments to pass to comparison functions and matching rule
-    kwargs: dict[typing.Any, typing.Any] = linkage_pass.get("kwargs", {})
+    matching_rule: matchers.MATCH_RULE_FUNC = utils.str_to_callable(algorithm_pass.rule)
```
Comment (ericbuckley) [inline src/recordlinker/linking/link.py:92]: ditto to my above comment on the evaluators

File: src/recordlinker/linking/mpi_service.py [hunk 3]
```diff
@@ -27,12 +27,10 @@ def get_block_data(
...
     # is considered a match.
-    for idx, block in enumerate(algo_config["blocks"]):
-        key_name = block["value"].upper()
-        if not hasattr(models.BlockingKey, key_name):
-            raise ValueError(f"Invalid Blocking Key: {block}")
-        # Get the matching Blocking Key based on the value in the algo_config
-        key = models.BlockingKey[key_name]
+    for idx, key_id in enumerate(algorithm_pass.blocking_keys):
+        #get the BlockingKey obj from the id
+        key = models.BlockingKey.from_id(key_id)
```
Comment (ericbuckley) [inline src/recordlinker/linking/mpi_service.py:32]: I think the `if not hasattr(models.BlockingKey, key_name)` check is still a good one here.  Because the data is coming from the model, we should be verifying that the data is valid before persisting now, so at this point in time it really can't be a user error.  But its possible that data is corrupted or some bug is introduced, so an assertion would be helpful here just in case.
```suggestion
        assert hasattr(models.BlockingKey, key_id), "Invalid Blocking Key: {key_id}")
        key = models.BlockingKey.from_id(key_id)
```

File: src/recordlinker/models/algorithm.py [hunk 4]
```diff
@@ -54,7 +54,7 @@ class AlgorithmPass(Base):
     algorithm_id: orm.Mapped[int] = orm.mapped_column(schema.ForeignKey("algorithm.id"))
     algorithm: orm.Mapped["Algorithm"] = orm.relationship(back_populates="passes")
     blocking_keys: orm.Mapped[list[int]] = orm.mapped_column(sqltypes.JSON)
-    evaluators: orm.Mapped[list[str]] = orm.mapped_column(sqltypes.JSON)
+    evaluators: orm.Mapped[dict] = orm.mapped_column(sqltypes.JSON)
```
Comment (ericbuckley) [inline src/recordlinker/models/algorithm.py:57]: why the change from `list[str]` to `dict`?

File: tests/unit/conftest.py [hunk 5]
```diff
@@ -26,12 +27,120 @@ def client():
...
 
-
 @pytest.fixture
 def basic_algorithm():
-    return utils.read_json_from_assets("linking", "basic_algorithm.json")["algorithm"]
-
+    basic_algo_pass1 = models.AlgorithmPass(
+        id=1, 
+        algorithm_id=1, 
+        blocking_keys=[1,2,3],
```
Comment (ericbuckley) [inline tests/unit/conftest.py:35]: I was thinking we'd store the actual names here of the Blocking Keys for readability sake.  It adds to the storage, but this table is going to be small.  For example

```
blocking_keys=["BIRTHDATE", "SEX", "FIRST_NAME"]
```

File: src/recordlinker/models/algorithm.py (hunk 4 above)
Comment (cbrinson-rise8) [reply src/recordlinker/models/algorithm.py:57]: Becuase it should have been a `dict` from the start. `evaluators` in the old `DIBBS_BASIC` and `DIBBS_ADVANCED` `algo_config`s where just `dict`.

File: src/recordlinker/linking/mpi_service.py (hunk 3 above)
Comment (cbrinson-rise8) [reply src/recordlinker/linking/mpi_service.py:32]: I added a `ValueError` in the `models.BlockingKey.from_id(...)` function that raises if there's no key with the given `key_id`. I was considering adding this assert but I don't think it will recognize the `key_id` as a attribute since its just an `int` and not an actual `BlockingKey` object

File: tests/unit/conftest.py (hunk 5 above)
Comment (cbrinson-rise8) [reply tests/unit/conftest.py:35]: I could make that change. I would just need to change the `AlgorithmPass` model to store a `list[str]` for `blocking_keys`

File: src/recordlinker/linking/mpi_service.py (hunk 3 above)
Comment (cbrinson-rise8) [reply src/recordlinker/linking/mpi_service.py:32]: I can go ahead and change `AlgorithmPass` model to store a `list[str]` like you commented about below, then I could remove the `from_id(...)` function and just do the assert statement you suggested above

File: src/recordlinker/linking/link.py (hunk 1 above)
Comment (cbrinson-rise8) [reply src/recordlinker/linking/link.py:90]: I was thinking about doing this but the `utils.bind_functions` takes in a `dict` and modifies the function `str` to a `callable` I put it here to avoid using a dict and keep `AlgorithmPass` everywhere else.

File: src/recordlinker/models/algorithm.py (hunk 4 above)
Comment (ericbuckley) [reply src/recordlinker/models/algorithm.py:57]: Ah thats right, thanks for the reminder I was confusing this with blocking_keys for a min.  Would you mind making this small change though?
```suggestion
    evaluators: orm.Mapped[dict[str,str]] = orm.mapped_column(sqltypes.JSON)
```

File: src/recordlinker/linking/link.py (hunk 1 above)
Comment (ericbuckley) [reply src/recordlinker/linking/link.py:90]: Ah, that's a great point, if we modify it it's really no longer a AlgorithmPass object to pass around.  I think there is room here for a future optimization with the Algorithm schema class I'm working on, but no need to hold this up for that.  Would you mind just adding a TODO comment here explaining the opportunity for an optimization if we can bind earlier in the stack, then I think we'll be good.

**Review by ericbuckley — APPROVED** [review]
Looks good @cbrinson-rise8 thanks for addressing some old mypy issues in main.py
